File: BST/4. Construct a balanced BST from the given keys/c++/code.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Node
{
public:
    int value;
    Node *left, *right;
    Node(int key, Node *left_item = NULL, Node *right_item = NULL) : value(key), left(left_item), right(right_item) {}
};

void ascending(Node *root)
{
    if (root == NULL)
    {
        return;
    }
    ascending(root->left);
    cout << root->value << " ";
    ascending(root->right);
}

Node *construct(vector<int> keys, int low, int high, Node *root)
{
    if (low > high)
    {
        return root;
    }

    int mid = (low + high) / 2;

    root = new Node(keys[mid]);

    root->left = construct(keys, low, mid - 1, root->left);
    root->right = construct(keys, mid + 1, high, root->right);

    return root;
}

int main()
{
    Node *root = NULL;

    vector<int> keys = {15, 10, 20, 8, 12, 16, 25};
    sort(keys.begin(), keys.end());

    root = construct(keys, 0, keys.size() - 1, root);
    ascending(root);
    cout << endl;
    return 0;
}

File: BST/5. Determine whether a given binary tree is a BST or not/python/code.py
class Node:
    left = None
    right = None
    parent = None

    def __init__(self,key):
        self.value = key

def insert(root, key):
    if root is None:
        return Node(key)

    if (key < root.value):
        root.left = insert(root.left, key)
    else:
        root.right = insert(root.right, key)
    
    return root

def checker(root, low, high):
    if root is None:
        return True
    if (root.value < low) or (root.value > high):
        return False
    
    return (checker(root.left, low, root.value - 1) and checker(root.right, root.value + 1, high))

root = None
arr = [] 
keys = [15, 10, 20, 8, 12, 16, 25]

for key in keys:
    root = insert(root,key)

if (checker(root, min(keys), max(keys))):
    print("Positive: Given binary tree is a BST")
else:
    print("Negative: Given binary tree is not a BST")